### PR TITLE
New Rule T0006: Avoid Helper in a type name

### LINKED ISSUE #9042: New Rule T0006: Avoid Helper in a type name

### PR DESCRIPTION
**Pull request by sebastien-marichal**
Fixes #9042

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@
 
 namespace SonarAnalyzer.Extensions
 {
-    internal static partial class SyntaxNodeExtensions
+    public static partial class SyntaxNodeExtensions
```
Comment (sebastien-marichal) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:25]: I had to make it public so I could use it in my rule.
Is there a better approach? Maybe `InternalVisibleTo`?

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:25]: Public is better than internals visible to

File: analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/HelperInTypeName.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class HelperInTypeName : StylingAnalyzer
+{
+    public HelperInTypeName() : base("T0006", "Do not use 'Helper' in type names.") { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+        {
+            var typeNameIdentifier = c.Node.GetIdentifier();
+            if (typeNameIdentifier?.ValueText.IndexOf("Helper", StringComparison.Ordinal) >= 0)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/HelperInTypeName.cs:32]: Use `.Contains`. It's ordinal, case sensitive and culture invariant.

File: analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/HelperInTypeName.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+        {
+            if (c.Node.GetIdentifier() is {} identifier && identifier.ValueText.Contains("Helper"))
+            {
+                c.ReportIssue(Rule, identifier);
+            }
+        },
+        TypeNameSyntaxKinds);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp.Styling/Rules/HelperInTypeName.cs:47]: We never do this. Put the list directly here, each item on a new line.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
The SyntaxKinds are overshooted

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**